python_modules/dagster/dagster/_core/definitions/sensor_definition.py

```diff
@@ -154,6 +160,205 @@ def repository_name(self) -> Optional[str]:
+        instance (Optional[DagsterInstance]): The deserialized instance can also be passed in
+            directly (primarily useful in testing contexts).
+
+    Example:
+
+    .. code-block:: python
+
+        from dagster import multi_asset_sensor, MultiAssetSensorEvaluationContext
+
+        @multi_asset_sensor(asset_keys=[AssetKey("asset_1), AssetKey("asset_2)])
+        def the_sensor(context: MultiAssetSensorEvaluationContext):
+            ...
+
+    """
+
+    def __init__(
+        self,
+        instance_ref: Optional[InstanceRef],
+        last_completion_time: Optional[float],
+        last_run_key: Optional[str],
+        cursor: Optional[Mapping[AssetKey, int]],
+        repository_name: Optional[str],
+        asset_keys: Sequence[AssetKey],
+        instance: Optional[DagsterInstance] = None,
+    ):
+        self._asset_keys = asset_keys
+        self.cursor_has_been_updated = False
+
+        self._exit_stack = ExitStack()
+        self._instance_ref = check.opt_inst_param(instance_ref, "instance_ref", InstanceRef)
+        self._last_completion_time = check.opt_float_param(
+            last_completion_time, "last_completion_time"
+        )
+        self._last_run_key = check.opt_str_param(last_run_key, "last_run_key")
+        self._cursor = check.opt_dict_param(cursor, "cursor", key_type=AssetKey)
+        self._repository_name = check.opt_str_param(repository_name, "repository_name")
+        self._instance = check.opt_inst_param(instance, "instance", DagsterInstance)
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, _exception_type, _exception_value, _traceback):
+        self._exit_stack.close()
+
+    @public  # type: ignore
+    @property
+    def instance(self) -> DagsterInstance:
+        # self._instance_ref should only ever be None when this MultiAssetSensorEvaluationContext was
+        # constructed under test.
+        if not self._instance:
+            if not self._instance_ref:
+                raise DagsterInvariantViolationError(
+                    "Attempted to initialize dagster instance, but no instance reference was provided."
+                )
+            self._instance = self._exit_stack.enter_context(
+                DagsterInstance.from_ref(self._instance_ref)
+            )
+        return cast(DagsterInstance, self._instance)
+
+    @public  # type: ignore
+    @property
+    def last_completion_time(self) -> Optional[float]:
+        return self._last_completion_time
+
+    @public  # type: ignore
+    @property
+    def last_run_key(self) -> Optional[str]:
+        return self._last_run_key
+
+    @public  # type: ignore
+    @property
+    def cursor(self) -> Optional[Mapping[AssetKey, Optional[int]]]:
+        """The cursor value for this sensor, which was set in an earlier sensor evaluation."""
+        return self._cursor
+
+    @public
+    def update_cursor(self, cursor: Optional[Mapping[AssetKey, Optional[int]]]) -> None:
```

Comment:
It seems a little odd to have both this and `advance_cursor`, given that they work very similarly and their names don't reflect the differences in how they work

Reply:
unresolving this because the cursor in this context is back to being stored as a string and it made more sense to have the context inherit from `SensorEvaluationContext` because of mypy. I was thinking i could update the two cursor functions to `update_cursor` and `update_all_cursors`, but overriding `update_cursor` with a new param name caused other type checking errors (and i thought was frowned upon in python but i could def be wrong about that)

So right now the state of the MultiAssetSensorEvaluationContext is:
`update_cursor`
`advance_cursor`
`advance_all_cursors`

I could override the body of `update_cursor` to throw an error, but that doesn't feel like a great solution either